beacon-chain/sync/rpc_beacon_blocks_by_range.go

```diff
@@ -214,6 +217,37 @@ func (s *Service) validateRangeRequest(r *pb.BeaconBlocksByRangeRequest) error {
 	return nil
 }
 
+// filters all the provided blocks to ensure they are canonical
+// and are strictly linear.
+func (s *Service) filterBlocks(ctx context.Context, blks []*ethpb.SignedBeaconBlock, roots [][32]byte, prevRoot *[32]byte,
+	step, startSlot uint64) ([]*ethpb.SignedBeaconBlock, error) {
+
+	newBlks := make([]*ethpb.SignedBeaconBlock, 0, len(blks))
+
+	for i, b := range blks {
+		isRequestedSlotStep := (b.Block.Slot-startSlot)%step == 0
+		isCanonical, err := s.chain.IsCanonical(ctx, roots[i])
+		if err != nil {
+			return nil, err
+		}
+		parentValid := *prevRoot != [32]byte{}
+		isLinear := *prevRoot == bytesutil.ToBytes32(b.Block.ParentRoot)
+		isSingular := step == 1
+		if isRequestedSlotStep && isCanonical {
+			// Exit early if our valid block is non linear.
```

Comment:
do we want to be nicer here and actually return the resorted blocks for linearity? or exit early is good enough

Reply:
We do return the blocks if we find the next block is non-linear. In the caller of this function, it checks the error and if it is invalidParent we just exit early and return what we have so far.